File: 15600/BOJ_15645.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    private static final int VALUE_SUM_LIMIT = 9*100000+1;
    private void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][][] dp = new int[n+1][3][2];
        for (int i = 1; i <= n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 3; j++) {
                int cur = Integer.parseInt(st.nextToken());
                dp[i][j][1] = VALUE_SUM_LIMIT;
                for (int k = j-1; k <= j+1; k++) {
                    if (k < 0 || k >= 3) continue;
                    dp[i][j][0] = Math.max(dp[i][j][0], dp[i-1][k][0]);
                    dp[i][j][1] = Math.min(dp[i][j][1], dp[i-1][k][1]);
                }
                dp[i][j][0]+=cur;
                dp[i][j][1]+=cur;
            }
        }
        int max = Math.max(dp[n][0][0], Math.max(dp[n][1][0], dp[n][2][0]));
        int min = Math.min(dp[n][0][1], Math.min(dp[n][1][1], dp[n][2][1]));
        System.out.println(max + " " + min);
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}

File: 15600/BOJ_15645.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

const val VALUE_SUM_LIMIT = 9*100000+1

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    var n = readLine().toInt()
    var max = Array(3){0}
    var min = Array(3){0}
    while(n-->0) {
        val st = StringTokenizer(readLine())
        val maxTmp = Array(3){0}
        val minTmp = Array(3){VALUE_SUM_LIMIT}
        var i = 0
        while (st.hasMoreTokens()) {
            val cur = st.nextToken().toInt()
            for (j in i-1..i+1) {
                if (j < 0 || j >= 3) continue
                maxTmp[i] = Math.max(maxTmp[i], max[j])
                minTmp[i] = Math.min(minTmp[i], min[j])
            }
            maxTmp[i]+=cur
            minTmp[i]+=cur
            i++
        }
        max = maxTmp
        min = minTmp
    }
    println("${max.toList().maxOrNull()} ${min.toList().minOrNull()}")
}
